File: src/net.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <sys/stat.h>
#include <curl/curl.h>
#include "bhpkg.h"

#ifndef SSL_CERT_PATH
#define SSL_CERT_PATH "/etc/ssl/certs/ca-certificates.crt"
#endif

struct XferInfo
{
  char name[128];
  double dl_total;
  double dl_now;
  char dest_path[PATH_MAX];
  char **mirrors;
  size_t mirror_count;
  size_t current_mirror;
  const char *expected_hash;
};

static void
parse_pipe_list (const char *str, char ***arr_out, size_t *count_out)
{
  char *copy, *tok;
  size_t count = 0;

  if (!str || !*str)
    {
      *count_out = 0;
      *arr_out = NULL;
      return;
    }
    
  copy = xstrdup (str);
  for (char *c = copy; *c; c++)
    if (*c == '|') count++;
    
  count++;
  
  *arr_out = xmalloc (count * sizeof (char *));
  tok = strtok (copy, "|");
  count = 0;
  
  while (tok)
    {
      while (*tok == ' ' || *tok == '\t') tok++;
      char *end = tok + strlen (tok) - 1;
      while (end > tok && (*end == ' ' || *end == '\t')) *end-- = '\0';
      
      (*arr_out)[count++] = xstrdup (tok);
      tok = strtok (NULL, "|");
    }
    
  *count_out = count;
  free (copy);
}

static int
progress_cb (void *p, curl_off_t dltotal, curl_off_t dlnow, curl_off_t ultotal, curl_off_t ulnow)
{
  struct XferInfo *info;

  (void) ultotal;
  (void) ulnow;

  if (UNLIKELY (g_interrupted))
    return 1;

  info = (struct XferInfo *) p;
  info->dl_total = (double) dltotal;
  info->dl_now = (double) dlnow;
  return 0;
}

int
net_download_all (BuildList *list)
{
  CURLM *multi;
  CURL **curls;
  FILE **files;
  struct XferInfo *infos;
  int running = 1;
  size_t total_downloads = 0;
  size_t idx = 0;
  bool download_failed = false;
  int last_percent = -1;
  int bar_width = 30;

  if (list->count == 0)
    return 0;

  for (size_t i = 0; i < list->count; i++)
    total_downloads += list->pkgs[i]->num_sources;

  if (total_downloads == 0)
    return 0;

  multi = curl_multi_init ();
  curl_multi_setopt (multi, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);

  curls = xmalloc (total_downloads * sizeof (CURL *));
  files = xmalloc (total_downloads * sizeof (FILE *));
  infos = xmalloc (total_downloads * sizeof (struct XferInfo));

  for (size_t i = 0; i < list->count; i++)
    {
      for (size_t j = 0; j < list->pkgs[i]->num_sources; j++)
        {
          char fn[PATH_MAX];
          snprintf (fn, sizeof (fn), "/var/lib/bhpkg/tmp/%s-%s-%zu.src", list->pkgs[i]->name, list->pkgs[i]->version, j);
          
          snprintf (infos[idx].name, sizeof (infos[idx].name), "%s", list->pkgs[i]->name);
          snprintf (infos[idx].dest_path, sizeof (infos[idx].dest_path), "%s", fn);
          infos[idx].dl_total = 0;
          infos[idx].dl_now = 0;          
          infos[idx].expected_hash = list->pkgs[i]->hashes ? list->pkgs[i]->hashes[j] : NULL;

          parse_pipe_list (list->pkgs[i]->sources[j], &infos[idx].mirrors, &infos[idx].mirror_count);
          infos[idx].current_mirror = 0;

          files[idx] = fopen (fn, "wbe");
          if (files[idx])
            fchmod (fileno (files[idx]), 0644);

          curls[idx] = curl_easy_init ();
          curl_easy_setopt (curls[idx], CURLOPT_URL, infos[idx].mirrors[0]);
          curl_easy_setopt (curls[idx], CURLOPT_WRITEDATA, files[idx]);
          curl_easy_setopt (curls[idx], CURLOPT_FOLLOWLOCATION, 1L);
          curl_easy_setopt (curls[idx], CURLOPT_MAXREDIRS, 10L);
          curl_easy_setopt (curls[idx], CURLOPT_FAILONERROR, 1L);
          curl_easy_setopt (curls[idx], CURLOPT_IPRESOLVE, CURL_IPRESOLVE_V4);
          curl_easy_setopt (curls[idx], CURLOPT_USERAGENT, "bhpkg/1.0 (Blackhole OS)");
          curl_easy_setopt (curls[idx], CURLOPT_XFERINFOFUNCTION, progress_cb);
          curl_easy_setopt (curls[idx], CURLOPT_XFERINFODATA, &infos[idx]);
          curl_easy_setopt (curls[idx], CURLOPT_NOPROGRESS, 0L);
          curl_easy_setopt (curls[idx], CURLOPT_CAINFO, SSL_CERT_PATH);
          curl_easy_setopt (curls[idx], CURLOPT_TCP_FASTOPEN, 1L);
          curl_easy_setopt (curls[idx], CURLOPT_TCP_KEEPALIVE, 1L);
          curl_easy_setopt (curls[idx], CURLOPT_BUFFERSIZE, 102400L);

          curl_multi_add_handle (multi, curls[idx]);
          idx++;
        }
    }

  if (g_verbosity >= 1)
    printf ("\033[?25l"); 

  while (running && !g_interrupted)
    {
      curl_multi_perform (multi, &running);
      curl_multi_poll (multi, NULL, 0, 100, NULL);

      double total_now = 0, total_expected = 0;
      for (size_t i = 0; i < total_downloads; i++)
        {
          total_now += infos[i].dl_now;
          if (infos[i].dl_total > 0)
            total_expected += infos[i].dl_total;
          else
            total_expected += infos[i].dl_now + 1;
        }

      double percent = (total_expected > 0) ? (total_now / total_expected) * 100.0 : 0.0;
      if (percent > 100.0) percent = 100.0;
      
      int current_percent = (int) percent;
      
      if (g_verbosity >= 1 && current_percent != last_percent)
        {
          int pos = (current_percent * bar_width) / 100;
          
          printf ("\r%s==>%s Fetching %zu sources... [", C_CYN, C_RST, total_downloads);
          if (g_pacman_mode)
            {
              for (int j = 0; j < bar_width; j++)
                {
                  if (j < pos) printf ("-");
                  else if (j == pos) printf ("%c", (current_percent % 2 == 0) ? 'C' : 'c');
                  else if (j % 2 == 0) printf ("o");
                  else printf (" ");
                }
            }
          else
            {
              for (int j = 0; j < bar_width; j++)
                {
                  if (j < pos) printf ("#");
                  else printf ("-");
                }
            }
          printf ("] %3.0f%%\033[K", percent);
          fflush (stdout);
          last_percent = current_percent;
        }
        
      int msgs_left;
      CURLMsg *msg;
      bool handle_retried = false;

      while ((msg = curl_multi_info_read (multi, &msgs_left))) 
        {
          if (msg->msg == CURLMSG_DONE) 
            {
              size_t i;
              for (i = 0; i < total_downloads; i++) 
                if (curls[i] == msg->easy_handle) break;

              bool success = (msg->data.result == CURLE_OK);

              if (success)
                {
                  if (files[i])
                    {
                      fclose (files[i]);
                      files[i] = NULL;
                    }

                  bool hash_ok = false;
                  
                  if (infos[i].expected_hash && infos[i].expected_hash[0] != '\0')
                    {
                      char *hash_copy = xstrdup (infos[i].expected_hash);
                      char *tok = strtok (hash_copy, "|");
                      while (tok)
                        {
                          while (*tok == ' ' || *tok == '\t') tok++;
                          char *end = tok + strlen (tok) - 1;
                          while (end > tok && (*end == ' ' || *end == '\t')) *end-- = '\0';

                          if (crypto_verify_sha256 (infos[i].dest_path, tok))
                            {
                              hash_ok = true;
                              break;
                            }
                          tok = strtok (NULL, "|");
                        }
                      free (hash_copy);
                    }
                  else
                    {
                      hash_ok = true; /* Skip verification for dynamic files without hashes */
                    }

                  if (!hash_ok)
                    {
                      if (g_verbosity >= 1) printf ("\r\033[K");
                      print_warn ("Hash mismatch for '%s' on %s. Suspected mirror desync.", 
                                  infos[i].name, infos[i].mirrors[infos[i].current_mirror]);
                      success = false; /* Force failover */
                    }
                }

              if (!success)
                {
                  if (infos[i].current_mirror + 1 < infos[i].mirror_count)
                    {
                      infos[i].current_mirror++;
                      if (g_verbosity >= 1) printf ("\r\033[K");
                      print_warn ("Falling back to: %s", infos[i].mirrors[infos[i].current_mirror]);
                      
                      curl_multi_remove_handle (multi, curls[i]);
                      if (files[i]) fclose (files[i]);
                      
                      files[i] = fopen (infos[i].dest_path, "wbe");
                      if (files[i]) fchmod (fileno (files[i]), 0644);
                      
                      curl_easy_setopt (curls[i], CURLOPT_URL, infos[i].mirrors[infos[i].current_mirror]);
                      curl_easy_setopt (curls[i], CURLOPT_WRITEDATA, files[i]);
                      curl_multi_add_handle (multi, curls[i]);
                      handle_retried = true;
                    }
                  else
                    {
                      if (g_verbosity >= 1) printf ("\r\033[K");
                      print_err ("All mirrors exhausted or failed integrity checks for '%s'.", infos[i].name);
                      if (msg->data.result != CURLE_OK)
                        print_err ("cURL Error: %s", curl_easy_strerror (msg->data.result));
                      download_failed = true;
                    }
                }
              else
                {
                   if (g_verbosity >= 1 && infos[i].expected_hash)
                     printf ("\r\033[K  %s[PASS]%s %s\n", C_GRN, C_RST, infos[i].name);
                }
            }
        }
      if (handle_retried) running = 1;
    }
    
  if (g_verbosity >= 1)
    {
      printf ("\r%s==>%s Fetching %zu sources... [", C_CYN, C_RST, total_downloads);
      for (int j = 0; j < bar_width; j++) printf (g_pacman_mode ? "-" : "#");
      printf ("] 100%%\033[K\n\033[?25h"); 
    }

  idx = 0;
  for (size_t i = 0; i < list->count; i++)
    {
      for (size_t j = 0; j < list->pkgs[i]->num_sources; j++)
        {
          curl_multi_remove_handle (multi, curls[idx]);
          curl_easy_cleanup (curls[idx]);
          if (files[idx])
            fclose (files[idx]);
          
          if (g_interrupted || download_failed)
            remove (infos[idx].dest_path);

          if (infos[idx].mirrors)
            {
              for (size_t k = 0; k < infos[idx].mirror_count; k++) free (infos[idx].mirrors[k]);
              free (infos[idx].mirrors);
            }
          idx++;
        }
    }

  curl_multi_cleanup (multi);
  free (curls);
  free (files);
  free (infos);

  return (g_interrupted || download_failed) ? -1 : 0;
}

File: src/net.rs
use crate::crypto;
use crate::error::{AppError, Result};
use crate::types::{AppContext, Package};
use curl::easy::Easy;
use curl::multi::Multi;
use std::collections::HashMap;
use std::io::Write;
use std::path::PathBuf;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::time::Duration;

struct DownloadTask {
    name: String,
    dest: PathBuf,
    mirrors: Vec<String>,
    current_mirror: usize,
    expected_hash: Option<String>,
    progress_idx: usize,
}

struct ActiveHandle {
    handle: curl::multi::EasyHandle,
    task: DownloadTask,
}

fn create_easy(
    task: &DownloadTask,
    progress_state: Arc<Vec<(AtomicU64, AtomicU64)>>,
    interrupted: Arc<AtomicBool>,
) -> Result<Easy> {
    let mut easy = Easy::new();

    let target_url = &task.mirrors[task.current_mirror];
    if !target_url.starts_with("http://") && !target_url.starts_with("https://") {
        return Err(AppError::Net(format!(
            "Insecure or unsupported protocol rejected: {}",
            target_url
        )));
    }
    easy.url(target_url)
        .map_err(|e| AppError::Net(e.to_string()))?;

    let mut options = std::fs::OpenOptions::new();
    options.write(true).create(true).truncate(true);
    use std::os::unix::fs::OpenOptionsExt;
    options.custom_flags(libc::O_NOFOLLOW);

    let mut file = match options.open(&task.dest) {
        Ok(f) => f,
        Err(_) => {
            let _ = std::fs::remove_file(&task.dest);
            options.open(&task.dest)?
        }
    };

    let interrupt_arc_write = Arc::clone(&interrupted);
    easy.write_function(move |data| {
        if interrupt_arc_write.load(Ordering::SeqCst) {
            return Ok(0);
        }
        if file.write_all(data).is_err() {
            return Ok(0);
        }
        Ok(data.len())
    })
    .map_err(|e| AppError::Net(e.to_string()))?;

    let interrupt_arc_hdr = Arc::clone(&interrupted);
    easy.header_function(move |header| {
        if let Ok(hdr_str) = std::str::from_utf8(header) {
            let hdr_lower = hdr_str.to_ascii_lowercase();
            if hdr_lower.starts_with("location:") {
                let loc = hdr_lower[9..].trim();
                if let Some(idx) = loc.find("://") {
                    let proto = &loc[..idx];
                    if proto != "http" && proto != "https" {
                        interrupt_arc_hdr.store(true, Ordering::SeqCst);
                        return false;
                    }
                }
            }
        }
        true
    })
    .map_err(|e| AppError::Net(e.to_string()))?;

    let state_arc = Arc::clone(&progress_state);
    let interrupt_arc_prog = Arc::clone(&interrupted);
    let idx = task.progress_idx;

    easy.progress_function(move |dltotal, dlnow, _, _| {
        if interrupt_arc_prog.load(Ordering::SeqCst) {
            return false;
        }
        state_arc[idx].1.store(dltotal.to_bits(), Ordering::Relaxed);
        state_arc[idx].0.store(dlnow.to_bits(), Ordering::Relaxed);
        true
    })
    .map_err(|e| AppError::Net(e.to_string()))?;

    easy.progress(true)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.follow_location(true)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.max_redirections(10)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.max_filesize(4_000_000_000)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.fail_on_error(true)
        .map_err(|e| AppError::Net(e.to_string()))?;

    easy.ssl_verify_peer(true)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.ssl_verify_host(true)
        .map_err(|e| AppError::Net(e.to_string()))?;

    let ca_path = std::path::Path::new("/etc/ssl/certs/ca-certificates.crt");
    if ca_path.exists() {
        easy.cainfo(ca_path)
            .map_err(|e| AppError::Net(e.to_string()))?;
    }

    easy.useragent("bhpkg/2.0 (Blackhole OS)")
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.buffer_size(102400)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.connect_timeout(Duration::from_secs(15))
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.low_speed_limit(100)
        .map_err(|e| AppError::Net(e.to_string()))?;
    easy.low_speed_time(Duration::from_secs(15))
        .map_err(|e| AppError::Net(e.to_string()))?;

    Ok(easy)
}

pub fn download_all(packages: &[Package], ctx: &AppContext) -> Result<()> {
    if packages.is_empty() {
        return Ok(());
    }

    let mut tasks_vec = Vec::new();
    let mut idx = 0;
    for pkg in packages {
        for (i, source) in pkg.sources.iter().enumerate() {
            let mirrors: Vec<String> = source.split('|').map(|s| s.trim().to_string()).collect();
            let dest = ctx.root_path(&format!(
                "var/lib/bhpkg/tmp/{}-{}-{}.src",
                pkg.name, pkg.version, i
            ));
            tasks_vec.push(DownloadTask {
                name: pkg.name.clone(),
                dest,
                mirrors,
                current_mirror: 0,
                expected_hash: pkg.hashes.get(i).cloned(),
                progress_idx: idx,
            });
            idx += 1;
        }
    }

    let total_tasks = tasks_vec.len();
    if total_tasks == 0 {
        return Ok(());
    }

    let progress_state = Arc::new(
        (0..total_tasks)
            .map(|_| (AtomicU64::new(0), AtomicU64::new(0)))
            .collect::<Vec<_>>(),
    );
    let is_done = Arc::new(AtomicBool::new(false));
    let interrupted = Arc::new(AtomicBool::new(false));

    let monitor_state = Arc::clone(&progress_state);
    let monitor_done = Arc::clone(&is_done);
    let monitor_ctx = ctx.clone();
    let pacman_mode = ctx.config.pacman_mode;

    let monitor_handle = std::thread::spawn(move || {
        let mut last_percent = -1;
        let bar_width = 30;

        if monitor_ctx.config.verbosity >= 1 {
            print!("\x1b[?25l");
        }

        while !monitor_done.load(Ordering::Relaxed) && !monitor_ctx.is_interrupted() {
            if monitor_ctx.config.verbosity >= 1 {
                let mut total_now = 0f64;
                let mut total_expected = 0f64;

                for state in monitor_state.iter() {
                    let now = f64::from_bits(state.0.load(Ordering::Relaxed));
                    let total = f64::from_bits(state.1.load(Ordering::Relaxed));
                    total_now += now;
                    if total > 0.0 {
                        total_expected += total;
                    } else {
                        total_expected += now + 1.0;
                    }
                }

                let mut percent = if total_expected > 0.0 {
                    (total_now / total_expected) * 100.0
                } else {
                    0.0
                };
                if percent > 100.0 {
                    percent = 100.0;
                }
                let current_percent = percent as i32;

                if current_percent != last_percent {
                    let pos = (current_percent * bar_width) / 100;
                    print!(
                        "\r\x1b[1;36m==>\x1b[0m Fetching {} sources... [",
                        total_tasks
                    );

                    if pacman_mode {
                        for j in 0..bar_width {
                            if j < pos {
                                print!("-");
                            } else if j == pos {
                                print!("{}", if current_percent % 2 == 0 { 'C' } else { 'c' });
                            } else if j % 2 == 0 {
                                print!("o");
                            } else {
                                print!(" ");
                            }
                        }
                    } else {
                        for j in 0..bar_width {
                            if j < pos {
                                print!("#");
                            } else {
                                print!("-");
                            }
                        }
                    }
                    print!("] {:3.0}%\x1b[K", percent);
                    std::io::stdout().flush().unwrap();
                    last_percent = current_percent;
                }
            }
            std::thread::sleep(Duration::from_millis(100));
        }

        if monitor_ctx.config.verbosity >= 1 {
            print!(
                "\r\x1b[1;36m==>\x1b[0m Fetching {} sources... [",
                total_tasks
            );
            for _ in 0..bar_width {
                print!("{}", if pacman_mode { "-" } else { "#" });
            }
            println!("] 100%\x1b[K\x1b[?25h");
        }
    });

    let mut multi = Multi::new();
    multi.pipelining(true, true).ok();

    let mut active_transfers: HashMap<usize, ActiveHandle> = HashMap::new();
    let mut pending_tasks = tasks_vec.into_iter();
    let mut overall_success = true;
    let max_concurrent = 16;

    let fill_queue = |multi: &mut Multi,
                      active: &mut HashMap<usize, ActiveHandle>,
                      pending: &mut std::vec::IntoIter<DownloadTask>,
                      success_flag: &mut bool| {
        while active.len() < max_concurrent {
            if let Some(task) = pending.next() {
                match create_easy(&task, Arc::clone(&progress_state), Arc::clone(&interrupted)) {
                    Ok(easy) => {
                        let token = task.progress_idx;
                        if let Ok(mut handle) = multi.add(easy) {
                            handle.set_token(token).unwrap();
                            active.insert(token, ActiveHandle { handle, task });
                        }
                    }
                    Err(e) => {
                        ctx.print_err(&format!("Failed to init download for {}: {}", task.name, e));
                        *success_flag = false;
                    }
                }
            } else {
                break;
            }
        }
    };

    fill_queue(
        &mut multi,
        &mut active_transfers,
        &mut pending_tasks,
        &mut overall_success,
    );

    while !active_transfers.is_empty() && !crate::SIGNAL_INTERRUPTED.load(Ordering::SeqCst) {
        multi.perform().map_err(|e| AppError::Net(e.to_string()))?;

        let mut done_tokens = Vec::new();
        multi.messages(|msg| {
            if let Some(result) = msg.result() {
                if let Ok(token) = msg.token() {
                    done_tokens.push((token, result));
                }
            }
        });

        for (token, result) in done_tokens {
            if let Some(mut active) = active_transfers.remove(&token) {
                let mut easy = multi.remove(active.handle).unwrap();

                let mut hash_ok = false;
                let mut err_msg = String::new();
                let mut http_code = 0;

                let success = match result {
                    Ok(_) => {
                        http_code = easy.response_code().unwrap_or(0);
                        http_code / 100 == 2
                    }
                    Err(e) => {
                        err_msg = format!("TLS/Network Error: {}", e);
                        false
                    }
                };

                drop(easy);

                if success {
                    if let Some(ref hash) = active.task.expected_hash {
                        if !hash.is_empty() {
                            hash_ok = crypto::verify_sha256(&active.task.dest, hash);
                            if !hash_ok {
                                err_msg = "SHA256 mismatch".into();
                            }
                        } else {
                            err_msg = "Expected hash is empty".into();
                        }
                    } else {
                        let file_name = active.task.dest.to_string_lossy();
                        if file_name.contains("repo_")
                            && (file_name.contains(".db-tmp-")
                                || file_name.contains(".db.sig-tmp-"))
                        {
                            hash_ok = true;
                        } else {
                            err_msg = "No hash provided for package".into();
                        }
                    }
                } else if err_msg.is_empty() {
                    err_msg = format!("HTTP {}", http_code);
                }

                if success && hash_ok {
                    if ctx.config.verbosity >= 1 && active.task.expected_hash.is_some() {
                        println!("\r\x1b[K  \x1b[1;32m[PASS]\x1b[0m {}", active.task.name);
                    }
                } else {
                    active.task.current_mirror += 1;
                    let _ = std::fs::remove_file(&active.task.dest);

                    if active.task.current_mirror < active.task.mirrors.len() {
                        ctx.print_warn(&format!(
                            "[{}] {} - Falling back to: {}",
                            active.task.name,
                            err_msg,
                            active.task.mirrors[active.task.current_mirror]
                        ));

                        interrupted.store(false, Ordering::SeqCst);
                        match create_easy(
                            &active.task,
                            Arc::clone(&progress_state),
                            Arc::clone(&interrupted),
                        ) {
                            Ok(new_easy) => {
                                let mut new_handle = multi.add(new_easy).unwrap();
                                new_handle.set_token(token).unwrap();
                                active_transfers.insert(
                                    token,
                                    ActiveHandle {
                                        handle: new_handle,
                                        task: active.task,
                                    },
                                );
                            }
                            Err(_) => {
                                overall_success = false;
                            }
                        }
                    } else {
                        ctx.print_err(&format!(
                            "All mirrors exhausted for '{}' (Last error: {})",
                            active.task.name, err_msg
                        ));
                        overall_success = false;
                    }
                }
            }
        }

        fill_queue(
            &mut multi,
            &mut active_transfers,
            &mut pending_tasks,
            &mut overall_success,
        );
        multi
            .wait(&mut [], Duration::from_millis(100))
            .map_err(|e| AppError::Net(e.to_string()))?;
    }

    if !overall_success || crate::SIGNAL_INTERRUPTED.load(Ordering::SeqCst) {
        interrupted.store(true, Ordering::SeqCst);
        for (_, active) in active_transfers {
            let _ = multi.remove(active.handle);
            let _ = std::fs::remove_file(&active.task.dest);
        }
        is_done.store(true, Ordering::Relaxed);
        let _ = monitor_handle.join();
        return Err(AppError::Net("Downloads failed or interrupted".into()));
    }

    is_done.store(true, Ordering::Relaxed);
    let _ = monitor_handle.join();

    Ok(())
}
